src/enzo/Grid_GrackleCustomCoolRate.C

```diff
@@ -0,0 +1,229 @@
+#include "typedefs.h"
+#include "global_data.h"
+#include "Fluxes.h"
+#include "GridList.h"
+#include "ExternalBoundary.h"
+#include "Grid.h"
+#include "fortran.def"
+#include "CosmologyParameters.h"
+ 
+/* This parameter controls whether the cooling function recomputes
+   the metal cooling rates.  It is reset by RadiationFieldUpdate. */
+ 
+extern int RadiationFieldRecomputeMetalRates;
+ 
+/* function prototypes */
+ 
+int CosmologyComputeExpansionFactor(FLOAT time, FLOAT *a, FLOAT *dadt);
+int GetUnits(float *DensityUnits, float *LengthUnits,
+	     float *TemperatureUnits, float *TimeUnits,
+	     float *VelocityUnits, FLOAT Time);
+int RadiationFieldCalculateRates(FLOAT Time);
+int FindField(int field, int farray[], int numfields);
+
+int grid::GrackleCustomCoolRate(int rank, int *dim, float *cool_rate,
+				float *dens, float *thrmeng,
+				float *velx, float *vely, float *velz,
+				float *HIdens, float *HIIdens,
+				float *HeIdens, float *HeIIdens, float *HeIIIdens,
+				float *edens,
+				float *HMdens, float *H2Idens, float *H2IIdens,
+				float *DIdens, float *DIIdens, float *HDIdens,
+				float *metaldens,
+				float *kphHI, float *kphHeI, float *kphHeII,
+				float *kdissH2I, float *gamma)
+{
+
+  // All passed fields should be in code units
+
+#ifdef USE_GRACKLE 
+
+  /* Return if this doesn't concern us. */
+
+  if (grackle_data->use_grackle == FALSE) return SUCCESS;
+  
+  if (RadiativeCooling == 0) return SUCCESS;
+ 
+  if (ProcessorNumber != MyProcessorNumber)
+    return SUCCESS;
+
+  int i;
+  int size = 1;
+  for (int d = 0; d < rank; d++)
+    size *= dim[d];
+ 
+  /* Compute the cooling time. */
+ 
+  FLOAT a = 1.0, dadt;
+  float TemperatureUnits = 1, DensityUnits = 1, LengthUnits = 1,
+    VelocityUnits = 1, TimeUnits = 1, aUnits = 1;
+
+  GetUnits(&DensityUnits, &LengthUnits, &TemperatureUnits,
+	   &TimeUnits, &VelocityUnits, Time);
+  if (ComovingCoordinates) {
+    CosmologyComputeExpansionFactor(Time+0.5*dtFixed, &a, &dadt);
+ 
+    aUnits = 1.0/(1.0 + InitialRedshift);
+  } else if (RadiationFieldRedshift > -1){
+    a       = 1.0 / (1.0 + RadiationFieldRedshift);
+    aUnits  = 1.0;
+  }
+  float afloat = float(a);
+
+  float *volumetric_heating_rate = NULL;
+  float *specific_heating_rate   = NULL;
+
+  // Double check if there's a metal field when we have metal cooling
+  int metal_cooling = MetalCooling;
+  if (metal_cooling && !metaldens) {
+    if (debug)
+      fprintf(stderr, "Warning: No metal field passed to GrackleCustomCoolRate. Not using metal cooling.\n");
+    metal_cooling = FALSE;
+  }
+
+  Eint32 *g_grid_dimension, *g_grid_start, *g_grid_end;
+  g_grid_dimension = new Eint32[3];
+  g_grid_start = new Eint32[3];
+  g_grid_end = new Eint32[3];
+
+  // Fortran code will act as if there are 3 dimensions regardless
+  for (i = 0; i < rank; i++) {
+    g_grid_dimension[i] = (Eint32) dim[i];
+    g_grid_start[i] = (Eint32) 0;
+    g_grid_end[i] = (Eint32) dim[i]-1;
+  }
+  for (i = rank; i < 3; i++){
```

Comment:
If you zero out g_grid_dimension, g_grid_start, and g_grid_end in this for loop (starting at line 114), why is it also set immediately above in line 109?  Am I missing something here?

Reply:
The two loops (line 109 and line 114) have different start and end values. The Fortran code called to calculate the cooling time will _always_ act as if the problem is three dimensional. So if you're calling this function for a 1D problem, the second loop sets entries corresponding to the unused second and third dimensions to zero. This is maybe a weird way of going about this, but I wanted this function to be written generally. I've updated the comments to hopefully make this clearer.

Reply:
whoops, I totally missed that.  Thank you!